unittests/FEXLinuxTests/tests/syscalls/syscalls_efault.cpp

```diff
@@ -0,0 +1,163 @@
+  auto ret = ::syscall(SYS_poll, invalid_fds, 1, 0);
+  REQUIRE(ret == -1);
+  CHECK(errno == EFAULT);
+}
+
+TEST_CASE("ppoll") {
+  // ppoll can return EFAULT for arguements 1, 3, 4
+  struct pollfd *invalid_fds = reinterpret_cast<struct pollfd *>(mmap(nullptr, sysconf(_SC_PAGESIZE), PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+  struct timespec *invalid_timespec = reinterpret_cast<struct timespec*>(mmap(nullptr, sysconf(_SC_PAGESIZE), PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+  sigset_t *invalid_sigset = reinterpret_cast<sigset_t*>(mmap(nullptr, sysconf(_SC_PAGESIZE), PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+
+  {
+    auto ret = ::syscall(SYS_ppoll, invalid_fds, 1, 0, nullptr, nullptr);
+    REQUIRE(ret == -1);
+    CHECK(errno == EFAULT);
+  }
+
+  {
+    struct pollfd valid_fds {
+      .fd = STDOUT_FILENO,
+      .events = 0,
+      .revents = 0,
+    };
+    auto ret = ::syscall(SYS_ppoll, &valid_fds, 1, invalid_timespec, nullptr, sizeof(uint64_t));
+    REQUIRE(ret == -1);
+    CHECK(errno == EFAULT);
+  }
+
+  {
+    struct pollfd valid_fds {
+      .fd = STDOUT_FILENO,
+      .events = 0,
+      .revents = 0,
+    };
+
+    struct timespec valid_ts{};
+    auto ret = ::syscall(SYS_ppoll, &valid_fds, 1, &valid_ts, invalid_sigset, sizeof(uint64_t));
+    REQUIRE(ret == -1);
+    CHECK(errno == EFAULT);
+  }
+
+  {
+    struct pollfd valid_fds {
+      .fd = STDOUT_FILENO,
+      .events = 0,
+      .revents = 0,
+    };
+
+    struct timespec valid_ts{};
+    sigset_t valid_sigset{};
+    sigemptyset(&valid_sigset);
+    auto ret = ::syscall(SYS_ppoll, &valid_fds, 1, &valid_ts, &valid_sigset, sizeof(uint64_t));
+    REQUIRE(ret == 0);
+  }
+
+  {
+    struct pollfd valid_fds {
+      .fd = STDOUT_FILENO,
+      .events = 0,
+      .revents = 0,
+    };
+
+    // Kernel will read timespec, but it then can't write the result back.
+    mprotect(invalid_timespec, sysconf(_SC_PAGESIZE), PROT_READ | PROT_WRITE);
+    invalid_timespec->tv_sec = 1;
+    mprotect(invalid_timespec, sysconf(_SC_PAGESIZE), PROT_READ);
+
+    sigset_t valid_sigset{};
+    sigemptyset(&valid_sigset);
+    auto ret = ::syscall(SYS_ppoll, &valid_fds, 1, invalid_timespec, &valid_sigset, sizeof(uint64_t));
+    REQUIRE(ret == 0);
+    CHECK(invalid_timespec->tv_sec == 1);
+  }
+}
+
+TEST_CASE("ppoll_64") {
+#ifndef SYS_ppoll_time64
+#define SYS_ppoll_time64 SYS_ppoll
+#endif
+  struct timespec64 {
+    uint64_t tv_sec, tv_nsec;
+  };
+
+  // ppoll can return EFAULT for arguements 1, 3, 4
+  struct pollfd *invalid_fds = reinterpret_cast<struct pollfd *>(mmap(nullptr, sysconf(_SC_PAGESIZE), PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+  timespec64 *invalid_timespec = reinterpret_cast<timespec64*>(mmap(nullptr, sysconf(_SC_PAGESIZE), PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+  sigset_t *invalid_sigset = reinterpret_cast<sigset_t*>(mmap(nullptr, sysconf(_SC_PAGESIZE), PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
```

Comment:
I'm guessing `mmap` is used here (instead of `new`/`unique_ptr`) to allow `mprotect` later in the test? Seems like a quirk that deserves a short comment.

Alternatively, making `invalid_timespec` a global `const` variable will also place it in read-only memory and allows us to use `unique_ptr` for everything else.

Reply:
mmap is used here to ensure we can fiddle with memory protections, which is why they all have PROT_NONE protections. And as you noted, so we can set a timeout and then set read-only for one of the tests.

Reply:
Moved the read-only timespec value to a const global and double-checked that it was in a RO data section and ensuring that it still got captured as expected.

Reply:
Added a comment.